### PR TITLE
Allow booleans to be lenient

### PR DESCRIPTION
**Pull request by colinmorelli**
This PR adds support for lenient booleans that accept `true` or `false` as strings (such as in query parameters) to be converted into booleans. Note that this _only_ allows exactly `true` or `false` - regardless of case. No other boolean-coercible values, such as 1 and 0, are allowed.

### CODE REVIEW

File: packages/codec/src/types/boolean.ts [hunk 1]
```diff
@@ -15,14 +20,25 @@ export class BooleanCodec extends BaseCodec<boolean> {
   }
 
   protected doDecode(value: unknown, context: Context): Validation<boolean> {
-    if (typeof value !== 'boolean') {
-      return failure(context, value, 'must be a boolean')
+    if (typeof value === 'boolean') {
+      return success(value)
+    } else if (typeof value === 'string') {
```
Comment (zia-r) [inline packages/codec/src/types/boolean.ts:25]: shouldn't it only check these if `this.isLenient`

File: packages/codec/src/types/boolean.spec.ts [hunk 2]
```diff
@@ -7,6 +7,18 @@ describe('Boolean Codec', () => {
 
   executeDecodeTests([
     decodeSuccessExpectation('parse simple boolean', codec, true, true),
+    decodeSuccessExpectation('parse lenient string', codec.lenient(), 'true', true),
```
Comment (zia-r) [inline packages/codec/src/types/boolean.spec.ts:10]: we want cases like 
`decodeFailureExpectation('not parse boolean string', codec, 'true', ['', 'must be a boolean']),` 
right?

Comment (colinmorelli) [reply packages/codec/src/types/boolean.spec.ts]: Yep. Added. Good spot.

File: packages/codec/src/types/boolean.ts (hunk 1 above)
Comment (colinmorelli) [reply packages/codec/src/types/boolean.ts]: ... 🤦 

Good catch. Updated.

File: packages/codec/src/types/number.ts [hunk 3]
```diff
@@ -40,7 +40,7 @@ export class NumberCodec extends BaseCodec<number> {
   }
 
   public lenient(): NumberCodec {
-    return new NumberCodec(true)
+    return new NumberCodec(true, this.options)
```
Comment (colinmorelli) [inline packages/codec/src/types/number.ts:43]: Just spotted this. Right now number codec only uses options to pass through documentation and examples so this wasn't caught in tests.

File: packages/codec/src/types/boolean.spec.ts [hunk 4]
```diff
@@ -7,6 +7,22 @@ describe('Boolean Codec', () => {
 
   executeDecodeTests([
     decodeSuccessExpectation('parse simple boolean', codec, true, true),
+    decodeSuccessExpectation('parse lenient string true', codec.lenient(), 'true', true),
+    decodeSuccessExpectation('parse lenient string false', codec.lenient(), 'false', false),
+    decodeSuccessExpectation('parse lenient string TRUE', codec.lenient(), 'TRUE', true),
+    decodeSuccessExpectation('parse lenient string FALSE', codec.lenient(), 'FALSE', false),
+    decodeFailureExpectation('not parse string when not in lenient mode', codec, 'true', [
```
Comment (zia-r) [inline packages/codec/src/types/boolean.spec.ts:14]: nit: for unit testing library code like this I personally would probably just duplicate all 4 test cases in lenient mode vs not

**Review by zia-r — APPROVED** [review]

**Pull request merged**